### PR TITLE
application: Matter bridge architectural refactoring

### PR DESCRIPTION
**Pull request by markaj-nordic**
Main scope:   
- allow to inject Matter and non-Matter devices to be bridgedinto the BridgeManager object
- added devices factory to support bridged devices creation
- implemented a dedicated map facilitating device object life time management
- added necessary BridgeManager modifications

    
Also:
- Added missing virtual constructor in abstract BridedDeviceDataProvider class.
- stop the timer in data provider when object is being destructed
- added more logs

I still see some things which can be improved/fixed, so all comments are warmly welcomed.

### CODE REVIEW

File: applications/matter_bridge/CMakeLists.txt [hunk 1]
```diff
@@ -34,6 +34,7 @@ target_include_directories(app PRIVATE
     src/bridged_device_types
     ${COMMON_ROOT}/src
     ${COMMON_ROOT}/src/bridge/
+    ${COMMON_ROOT}/src/bridge/util
```
Comment (kkasperczyk-no) [inline applications/matter_bridge/CMakeLists.txt:37]: I think that such directory doesn't exist.

File: applications/matter_bridge/src/bridged_device_factory.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include "temperature_sensor.h"
+#include "temperature_sensor_data_provider.h"
+#endif
+
+namespace BridgeFactory
+{
+
+using UpdateAttributeCallback = BridgedDeviceDataProvider::UpdateAttributeCallback;
+using DeviceType = BridgedDevice::DeviceType;
+using BridgeDeviceFactory = DeviceFactory<BridgedDevice, const char *>;
```
Comment (kkasperczyk-no) [inline applications/matter_bridge/src/bridged_device_factory.h]:
```suggestion
using BridgedDeviceFactory = DeviceFactory<BridgedDevice, const char *>;
```

File: applications/matter_bridge/src/bridged_device_factory.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+using DataProviderFactory = DeviceFactory<BridgedDeviceDataProvider, UpdateAttributeCallback>;
+
+inline BridgeDeviceFactory &GetBridgeDeviceFactory()
```
Comment (kkasperczyk-no) [inline applications/matter_bridge/src/bridged_device_factory.h]:
```suggestion
inline BridgedDeviceFactory &GetBridgedDeviceFactory()
```

File: applications/matter_bridge/CMakeLists.txt (hunk 1 above)
Comment (markaj-nordic) [reply applications/matter_bridge/CMakeLists.txt:37]: Right, it sneaked in during rebasing. Now it's added.

File: samples/matter/common/src/bridge/bridge_manager.cpp [hunk 4]
```diff
@@ -48,57 +48,15 @@ void BridgeManager::Init()
...
-	Platform::UniquePtr<BridgedDevice> device(nullptr);
-	Platform::UniquePtr<BridgedDeviceDataProvider> dataProvider(nullptr);
-
+	const char *nodeLabel = aDevice->GetNodeLabel();
 	if (nodeLabel && (strlen(nodeLabel) >= BridgedDevice::kNodeLabelSize)) {
 		return CHIP_ERROR_INVALID_STRING_LENGTH;
 	}
 
-	switch (bridgedDeviceType) {
-#if CONFIG_BRIDGE_ONOFF_LIGHT_BRIDGED_DEVICE
```
Comment (kkasperczyk-no) [inline samples/matter/common/src/bridge/bridge_manager.cpp]: We have similar ifdefs at the top of the file, so they are not needed anymore as well.

File: samples/matter/common/src/bridge/bridge_shell.cpp [hunk 5]
```diff
@@ -20,8 +21,17 @@ static int add_bridged_device(const struct shell *shell, size_t argc, char **arg
...
 
-	CHIP_ERROR err =
-		GetBridgeManager().AddBridgedDevice(static_cast<BridgedDevice::DeviceType>(deviceType), nodeLabel);
+	auto *newBridgedDevice = BridgeFactory::GetBridgedDeviceFactory().Create(
+		static_cast<BridgedDevice::DeviceType>(deviceType), nodeLabel);
+	auto *newDataProvider = BridgeFactory::GetDataProviderFactory().Create(
+		static_cast<BridgedDevice::DeviceType>(deviceType), BridgeManager::HandleUpdate);
+
+	if (!newBridgedDevice || !newDataProvider) {
+		shell_fprintf(shell, SHELL_INFO, "Unsupported device type\n");
```
Comment (kkasperczyk-no) [inline samples/matter/common/src/bridge/bridge_shell.cpp]: I think that we may also fall into this case, if the memory could not be allocated using new, right? Maybe we should try to handle errors through some error code or print the generic error message without giving any details about the reason.

File: samples/matter/common/src/bridge/bridge_manager.h [hunk 6]
```diff
@@ -6,35 +6,56 @@
...
+	static CHIP_ERROR HandleWrite(uint16_t index, chip::ClusterId clusterId,
 				      const EmberAfAttributeMetadata *attributeMetadata, uint8_t *buffer);
 	static void HandleUpdate(BridgedDeviceDataProvider &dataProvider, chip::ClusterId clusterId,
 				 chip::AttributeId attributeId, void *data, size_t dataSize);
 
 private:
-	CHIP_ERROR AddDevice(BridgedDevice *device);
-
 	static constexpr uint8_t kMaxBridgedDevices = CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT;
 	static constexpr uint8_t kMaxDataProviders = CONFIG_BRIDGE_MAX_BRIDGED_DEVICES_NUMBER;
```
Comment (kkasperczyk-no) [inline samples/matter/common/src/bridge/bridge_manager.h:28]: It seems it is not used anywhere. For now we create a new data provider for every bridged device, but it doesn't need to be this way always. I think that we should somehow validate that we will not create more providers than it's assumed.

File: samples/matter/common/src/bridge/bridge_shell.cpp [hunk 7]
```diff
@@ -20,8 +21,17 @@ static int add_bridged_device(const struct shell *shell, size_t argc, char **arg
 		nodeLabel = argv[2];
 	}
 
-	CHIP_ERROR err =
-		GetBridgeManager().AddBridgedDevice(static_cast<BridgedDevice::DeviceType>(deviceType), nodeLabel);
+	auto *newBridgedDevice = BridgeFactory::GetBridgedDeviceFactory().Create(
```
Comment (kkasperczyk-no) [inline samples/matter/common/src/bridge/bridge_shell.cpp]: I was wondering if we could improve this design even more. So far we've got `BridgeManager` that is independent from supported device types and we've got generic `bridge_util` module. However in the `bridge_shell` we use the `bridge_device_factory` that is specific implementation and may not be used by the user.

What do you think about adding `BridgedDeviceFactory` and `DataProviderFactory` as a class members of a `BridgeManager`. The members could be initialized in the `app_task.cpp` using the `bridged_device_factory` specific implementation. This way we would use the specific implementation only in the `app_task` and the rest of the code would be completely device-type agnostic. In the shell module we could get factories using some getter method of a `BridgeManager`.

File: samples/matter/common/src/bridge/bridge_shell.cpp (hunk 5 above)
Comment (markaj-nordic) [reply samples/matter/common/src/bridge/bridge_shell.cpp]: Right, additional error handling would complicate the API, so I will just modify the error message.

File: samples/matter/common/src/bridge/bridge_manager.h (hunk 6 above)
Comment (markaj-nordic) [reply samples/matter/common/src/bridge/bridge_manager.h:28]: Ok, I added a check for that in the BridgeManager.

File: samples/matter/common/src/bridge/bridge_shell.cpp (hunk 7 above)
Comment (markaj-nordic) [reply samples/matter/common/src/bridge/bridge_shell.cpp]: I would rather think of `BridgeFactory` as an independent module and avoid placing its objects inside any of the current classes. The idea was that the user was responsible for providing factories with desired device types, and the current implementation can be treated as a template which can be easily extended with further device types. 

The shell module uses what has been put in factories by default, but these can be used in the same manner in other places. If we decide that there might be multiple (customized) factories at once, then one can use `DataProviderFactory` and implement them (or maybe then, we should even consider dropping the idea of the factory per se). We may also rename the current ones as `GetDefaultBridgedDeviceFactory` and `GetDefaultDataProviderFactory`. 

We can discuss it f2f :)

Comment (kkasperczyk-no) [reply samples/matter/common/src/bridge/bridge_shell.cpp]: Ok, as discussed F2F, let it be in the shell for now.

**Review by kkasperczyk-no — APPROVED** [review]

**Comment by markaj-nordic** [comment]
PR moved to the nRF Connect SDK.

**Pull request closed without merging**